net, test: invalid p2p messages and test framework improvements

jonatack opened the pull request:
...seen while reviewing #19264, #19252, #19304 and #19107:

in `net_processing.cpp`
- make the debug logging for oversized message size misbehavior the same for `addr`, `getdata`, `headers` and `inv` messages

in `p2p_invalid_messages`
- add missing logging
- improve assertions/message sends, move cleanup disconnections outside the assertion scopes
- split a slowish 3-part test into 3 order-independent tests
- add a few p2p constants to the test framework

jonatack committed:
test: add p2p_invalid_messages logging

jonatack committed:
test: refactor test_large_inv() into 3 tests with common method

jonatack committed:
test: hoist p2p values to test framework constants

jonatack committed:
test: improve msg sends and p2p disconnections in p2p_invalid_messages

- call disconnect_p2ps() outside of the assert_debug_log scopes
- send messages directly from the p2p conn rather than via nodes[0].p2p
- add an assertion

MarcoFalke commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -93,40 +101,47 @@ def test_msgtype(self):
-            conn.send_and_ping(msg)
+    def test_oversized(self, msg_type, msg, size):
+        self.log.info("Test {} message of size {} is logged as misbehaving".format(msg_type, size))
+        with self.nodes[0].assert_debug_log(['Misbehaving', msg_type, str(size)]):
```
splitting up msg_type and size could match them on different lines, while we want to match both on the same line

jonatack replied:
Good point! Fixed. This made it apparent that in `net_processing.cpp` the debug logging for misbehaving from `headers` message size was a bit different from the equivalent misbehavior logging for `inv`, `getdata`, and `addr` message sizes. Fixed. If we don't want to touch it, we can also have each test pass in their own debug log message to test for.

troygiorshev commented:
ACK 6b18bfefed402a904db45a19b9cd8c1cb5d4c337

troygiorshev reviewed (approved):
Reviewed and ran all tests. Just one suggestion if you plan on making any changes.

jonatack commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -93,40 +101,47 @@ def test_msgtype(self):
 
         # Check that, even though the node is being hammered by nonsense from one
         # connection, it can still service other peers in a timely way.
+        self.log.info("(b) Check node still services peer in a timely way")
```
Done.

jonatack commented:
Thanks for reviewing @troygiorshev. Took your suggestion and also removed an unnecessary `disconnect_p2ps()` call in `test_oversized_msg()`.

gzhao408 commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -72,16 +78,17 @@ def test_checksum(self):
+        with self.nodes[0].assert_debug_log(['disconnecting peer=']):
             msg = msg_unrecognized(str_data="d" * (VALID_DATA_LIMIT + 1))
             msg = conn.build_message(msg)
             self.nodes[0].p2p.send_raw_message(msg)
```
Here 
```suggestion
            conn.send_raw_message(msg)
```

glozow reviewed (commented):
Concept ACK
Mentioned this elsewhere but my edit is already outdated. Small suggestion for p2p_invalid_messages.py, feel free to ignore: 
If you replace the 3 (specifically, `test_checksum`, `test_size`, and `test_msgtype`)  `self.nodes[0].p2p` with `conn`, all of the subtest `disconnect_p2ps` can also be removed. The only reason they need it now is to clear `p2ps`.

jonatack committed:
net: update misbehavior logging for oversized messages

so that oversized ADDR, GETDATA, HEADERS and INV messages print
the same consistent debug logs.

jonatack commented:
Thank you for reviewing, @gzhao408. Took your suggestion and credited you as co-author in the commit.

jnewbery commented:
I don't think we should be removing the `disconnect_p2ps` calls at the end of each subtest. Ideally the subtest finishes by leaving the node and fixture in as close a state to they were in originally. See https://github.com/bitcoin/bitcoin/pull/19304#discussion_r441643412

gzhao408 commented:
@jnewbery good point... 
> Ideally the subtest finishes by leaving the node and fixture in as close a state to they were in originally

In terms of state, I'm thinking state = (1) the `test_node.p2ps` array and (2) the connections between the node and its peers. 

(1) I think it's ok if the array has an extra entry if we're not using the `.p2p` property? Would you agree? If not, then yeah, we need to put `disconnect_p2ps()` back 😄 .

(2) For connections: some of the subtests call `wait_for_disconnect` which makes sure that they are disconnected from the peer's POV (no peer._transport). With #19252,`disconnect_p2ps` waits for node's POV (not in node.getpeerinfo()). 

I'm not sure if disconnected from peer's POV  == disconnected from the node's POV. If they're effectively the same, I'd say it's unnecessary to call `disconnect_p2ps` after `wait_for_disconnect`, and it could be cleaner to call `conn.peer_disconnect` for the ones that aren't connected at the end of the subtest.

I might be overthinking this haha, I dug into this when I thought superfluous `disconnect_p2ps` would cause performance problems 😅 .

jonatack commented:
AFAICT the tests are order-independent with this PR, and I checked that this passes when rebased to current master:
```diff
+from test_framework.util import assert_equal
  
+    def assert_no_connected_peers(self):
+        assert_equal(self.nodes[0].num_connected_mininodes(), 0)
+
     def run_test(self):
         self.test_magic_bytes()
+        self.assert_no_connected_peers()
         self.test_checksum()
+        self.assert_no_connected_peers()
         self.test_size()
+        self.assert_no_connected_peers()
         self.test_msgtype()
+        self.assert_no_connected_peers()
         self.test_oversized_inv_msg()
+        self.assert_no_connected_peers()
         self.test_oversized_getdata_msg()
+        self.assert_no_connected_peers()
         self.test_oversized_headers_msg()
+        self.assert_no_connected_peers()
         self.test_resource_exhaustion()
+        self.assert_no_connected_peers()
```

jonatack commented:
Anyway, removed the removal of `disconnect_p2ps`, and rebased to master to be sure all is well with the latest version of `disconnect_p2ps`.

jonatack commented:
@MarcoFalke several PRs on this test have been merged but no approval from you on this one. I think they are good improvements but feel free to close it if you disagree; it will save me from spinning my wheels.

MarcoFalke commented:
I generally use the conflicts list of DrahtBot to see which pr has the most reviews and is the most ready to merge and then get them in one by one if they have received their ACKs.

Will take another look here. One sec

MarcoFalke commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -39,19 +43,22 @@ def __repr__(self):
     def set_test_params(self):
         self.num_nodes = 1
-        self.setup_clean_chain = True
+        self.setup_clean_chain = False
```
why is this changed?

jonatack replied:
i thought it was faster but can revert that change while rebasing

MarcoFalke replied:
why would it be faster to copy the blocksdir as opposed to copy nothing?

jonatack replied:
I (mis)understood(?) the discussions to say that using the cached chain was faster than spinning up a new one. Reverted.

MarcoFalke replied:
There is no new chain generated in this test:

```
$ git grep generate 'test/functional/p2p_invalid_messages.py'
$

MarcoFalke commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -19,7 +22,8 @@
 from test_framework.test_framework import BitcoinTestFramework
 
 MSG_LIMIT = 4 * 1000 * 1000  # 4MB, per MAX_PROTOCOL_MESSAGE_LENGTH
```
```suggestion
MSG_LIMIT = MAX_PROTOCOL_MESSAGE_LENGTH
```

jonatack replied:
Good eye; that's an error introduced this morning in my last rebase, will fix

jonatack replied:
That line should be removed.

jonatack replied:
Removed the line.

maflcko reviewed (commented):
Fine with me

jonatack commented:
Rebased for merged #19304, reverted the chain setting, and removed a line left over from the last change.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3245,7 +3245,7 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
         if (nCount > MAX_HEADERS_RESULTS) {
             LOCK(cs_main);
-            Misbehaving(pfrom.GetId(), 20, strprintf("headers message size = %u", nCount));
+            Misbehaving(pfrom.GetId(), 20, strprintf("message headers size() = %u", nCount));
```
I don't understand this change. This is a "headers" message, so the original log seems correct - that it's a headers message of size x.

jonatack replied:
I don't disagree but opted for the smallest change. Do you think the `inv`, `getdata`, and `addr` oversized messages misbehavior logging should be changed instead? Or are they all good as-is (in which case the `headers` one remains different, and each specific one can be passed into the `test_oversized_msg()` assertion function from the caller.)

jnewbery replied:
ah ok, it's to be consistent with the other logs. That's fine.

troygiorshev replied:
IMO they should all be changed to `<type> message size = %u`.  "`<type> message`" because that sounds better to my ear, and "`size`" as opposed to "`size()`" because we're talking about the value of the function, not the function itself.

That said, I'll ACK anything that's consistent among these 4 logs.

jonatack replied:
Done.

jnewbery commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -86,73 +92,83 @@ def test_magic_bytes(self):
-        self.log.info("Sending a bunch of large, junk messages to test memory exhaustion. May take a bit...")
-
-        # Run a bunch of times to test for memory exhaustion.
+        self.log.info("(a) Send 80 messages of max data size 4MB to test for memory exhaustion")
```
This comment "test for memory exhaustion" is no longer relevant. When the test was written, it tested the node's memory usage before and after sending the messages and asserted that it hadn't gone up. That was an unreliable test and so the memory checking was removed.

jonatack replied:
Thanks @jnewbery, what change would you suggest instead?

jonatack replied:
We could just remove "to test for memory exhaustion" but saying why could be good here.

jnewbery replied:
Yes, I suggest removing "to test for memory exhaustion" because the test isn't doing that.

jonatack replied:
Thanks -- done.

jnewbery commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -24,8 +27,8 @@
-MSG_LIMIT = 4 * 1000 * 1000  # 4MB, per MAX_PROTOCOL_MESSAGE_LENGTH
-VALID_DATA_LIMIT = MSG_LIMIT - 5  # Account for the 5-byte length prefix
+VALID_DATA_LIMIT = MAX_PROTOCOL_MESSAGE_LENGTH - 5  # Account for the 5-byte length prefix
+DISCONNECTING = 'disconnecting peer='
```
My preference is not to test the debug log for actions that we can test directly. Here, we can directly test that the peer has been disconnected, so I don't think we need to test for this message in the debug log.

jonatack replied:
I added this after noticing both here and in the #19107 that some of the tests aren't testing for the disconnection, and it's a free additional check. If a change made one of these disconnections no longer be logged without warning and no test needing to be changed, ISTM that could be a regression.

jnewbery replied:
Right - I think the fix is to test for disconnection, not to test for the log message that accompanies disconnection.

jonatack replied:
Ok, in three cases it was only logging the disconnection because `disconnect_p2ps()` was previously being called in the scope of the `assert_debug_log`. Updating along with my review feedback in #19107.

jonatack replied:
Done -- good call, thanks.

jnewbery reviewed (commented):
Nice changes, Jon. I've left a few minor comments inline.

troygiorshev commented:
reACK ca658361c7e6a137d1ee8be8a005b35ad0426fa3

Reviewed and stepped through each test.  Verified that none of these changes caused linting errors.

maflcko reviewed (changes requested):
c6e7519647ef8736aaa5d8e7f804d0340a36029a

jonatack commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -80,49 +80,49 @@ def test_buffer(self):
         self.log.info("Test message with invalid magic bytes disconnects peer")
         conn = self.nodes[0].add_p2p_connection(P2PDataStore())
-        with self.nodes[0].assert_debug_log(['PROCESSMESSAGE: INVALID MESSAGESTART badmsg']):
+        with self.nodes[0].assert_debug_log('PROCESSMESSAGE: INVALID MESSAGESTART badmsg'):
```
Thanks, done. I should have verified the method. Am in the long-standing habit of accepting either a string or an array in framework APIs I make, e.g.
```diff
     def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
+        if type(expected_msgs) == str:
+            expected_msgs = [expected_msgs]
+        if type(unexpected_msgs) == str:
+            unexpected_msgs = [unexpected_msgs]
         if unexpected_msgs is None:
             unexpected_msgs = []
```

troygiorshev replied:
plug: https://github.com/bitcoin/bitcoin/pull/18210

MarcoFalke replied:
I did try to add `: List[str]`, but for some reason it didn't complain when I passed a string. Maybe mypy was broken on my machine?

jonatack commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -122,20 +124,21 @@ def test_msgtype(self):
-        with self.nodes[0].assert_debug_log(['Misbehaving', '(40 -> 60): headers message size = 2001']):
-            msg = msg_headers([CBlockHeader()] * 2001)
-            conn.send_and_ping(msg)
+    def test_oversized_msg(self, msg_type, msg, size):
```
Nice; done.

troygiorshev commented:
reACK 98af0f02055409af7f3d1a5df06d315f36b0b856

jonatack commented:
Rebased, no changes. Thank you for ACKing @troygiorshev! Should be trivial to re-ack with `git range-diff 8ef15e8 98af0f0 56010f9`

troygiorshev commented:
reACK 56010f92564a94b0ca6c008c0e6f74a19fad4a2a

rajarshimaitra reviewed (commented):
Concept ACK. Nice changes, easy to review. Compiled without issue, verified all unit and functions tests passing.
Below are a nonblocking nit and a question for explanation.

MarcoFalke commented:
ACK 56010f9256 🎛

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 56010f9256 🎛
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUicWwv+KIsi9o22/6uZbUTRM0UsyIPesQTV0ZcUzpXX8Sz27WY/psDv60xlD3+W
eUGgldLLYL1EBH8wWRbyMWsQ4WFQJILWXl6+8Cr2b8Lv9N1NhJ/z+u+FLfHHMlXt
dRxOMXXsytLgH8g3eW8f7zWbH0RaHtre+gcBiYqWAEbzq4i8D8h4lEO8ggA5c9TC
iGgj4ACk04YvOXDTOOgfF4DJQXIn3CuHrh7y55zo5X7+hZp43YEDRaVtKnggLTiM
dXZ3r3p+6i4Wqn2uFiZuw06rJN6PtSPEcBa8T3dmGVHixzPlXV5W3aILjU9TUGSV
tMoysA6J0N/RPZVEL0pucUOjrJPB3l6ULSEnl4anjC9rvURvaWJogqwJ5Fa/73Tq
OAFMkcO5iZod95zCDr6Alt34VI7nsJ//rej6/qpkeOSNc6vhQFbuSPaj5QNo/2Sc
BF2ySsHCMluT0JrkGY8EYvMUJka5cO7D894tRvaudfhn6wWT2mNc/E+95LTsGflG
EWs3YKPd
=Yt/T
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `937d2e5bea5f5ff38ae343e4664be89319c7ac06458b75aca1a579bbd3abbcd1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108937d2e5bea5f5ff38ae343e4664be89319c7ac06458b75aca1a579bbd3abbcd1f0102b1b41f1e5f0b64c1d09c95363978a9908fff01045f90d52f670c538a5bda1949651430208f1204bf312dcf8baa34258f9e9b8005e1055b20f6470a242c56ad7f88bc3a186139b08f1045ef3af44f0089f54ff4fefbb29080083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0104c45f01c2e4bb0ad5eb0f01c80a7981008f0208e3707d6e366ac7eb1843df3600434f5f390219b1a4c15483245c7f2691c9c8308f1045ef3af43f008d797f720ecc915180083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010b6b9f4ae5b94e8db586f114c63b4df0308f020a50b3180783133877baa12ebfcac622aa6e763393fccf5a56edc6df9ef3ca0f608f1045ef3af43f0083e890a2a6bcd23fa0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jonatack commented on test/functional/p2p_invalid_messages.py:
```diff
@@ -24,8 +27,7 @@
 
-MSG_LIMIT = 4 * 1000 * 1000  # 4MB, per MAX_PROTOCOL_MESSAGE_LENGTH
-VALID_DATA_LIMIT = MSG_LIMIT - 5  # Account for the 5-byte length prefix
+VALID_DATA_LIMIT = MAX_PROTOCOL_MESSAGE_LENGTH - 5  # Account for the 5-byte length prefix
```
We set `VALID_DATA_LIMIT` so that the length of the serialised `msg_unrecognized` message built with it will equal `MAX_PROTOCOL_MESSAGE_LENGTH`. This is verified at line 153; you can also add the following assertion at line 111
```diff
             msg = msg_unrecognized(str_data="d" * (VALID_DATA_LIMIT + 1))
+            assert_equal(len(msg.serialize()), MAX_PROTOCOL_MESSAGE_LENGTH + 1)
```
or try changing the offset from 5 to 4 or 6.

I've added this assertion and some other ones while testing @troygiorshev's net message headers refactoring and may propose the changes if they look worthwhile to me after a few days' time.
